stream/stream.go

```diff
@@ -0,0 +1,194 @@
+// returns it.
+func getCapability(req *chirp.Request) (string, error) {
+	if len(req.Data) < capabilityLen {
+		return "", errors.New("payload too short")
+	}
+	ret := string(req.Data[len(req.Data)-capabilityLen:])
+	// Trim the slice capacity so that a cheeky peer can't just grow
+	// the slice and recover the capability.
+	req.Data = slices.Clip(req.Data[:len(req.Data)-capabilityLen])
+	return ret, nil
+}
+
+// Call sends a call to the remote peer for the specified method and
+// data, and yields a stream of responses. The response stream ends at
+// the peer's discretion, or when ctx is canceled.
+//
+// The returned iterator yields zero or more (bs, nil) values. If the
+// call ends unsuccessfully, the iterator ends the stream with a final
+// (nil, err) tuple.
+func Call(ctx context.Context, peer *chirp.Peer, method string, req []byte) iter.Seq2[[]byte, error] {
+	return func(yield func([]byte, error) bool) {
+		capability := mkCapability()
+
+		req = append(req, capability...)
+
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		// The peer streams values back to us by calling the passed-in
+		// capability, which will run this handler in a different
+		// goroutine. We're in an iterator func, we can't yield from a
+		// random other goroutine. So, make a channel to smuggle
+		// payloads from the capability handler back to us for
+		// yielding.
+		vals := make(chan []byte)
+		peer.Handle(capability, func(callbackCtx context.Context, req *chirp.Request) ([]byte, error) {
+			select {
+			case vals <- req.Data:
+				return nil, nil
+			case <-ctx.Done():
+				// Client side cancellation, we're already unwinding
+				// this web of RPCs and just need the server to get on
+				// with that.
+				//
+				// This is also important to unwedge a buggy server
+				// that somehow manages to smuggle the capability out
+				// of the streaming response lifecycle, and tries to
+				// call us after the peer.Call() below has
+				// returned. This path serves as "what are you doing
+				// here? Go away" in the window between the call
+				// returning and the handler registration being
+				// cleaned up.
+				return nil, ctx.Err()
+			case <-callbackCtx.Done():
+				// Server cancellation (or maybe indirectly client
+				// cancellation too), we'll be told why when our Call
+				// below returns.
+				return nil, callbackCtx.Err()
+			}
+		})
+
+		errch := make(chan error, 1)
+		go func() {
+			// Note, unregister the capability handler in this
+			// goroutine, not the calling context, so that the other
+			// side can't hit "method not found" as the iterator shuts
+			// down.
+			defer peer.Handle(capability, nil)
+			defer close(errch)
+			_, err := peer.Call(ctx, method, req)
+			if ctx.Err() != nil {
+				// This codepath isn't strictly necessary, but it
+				// makes tests deterministic: a client-side
+				// cancellation can manifest in two ways: the
+				// peer.Call() notices first and returns ctx.Err()
```

Comment:
FWIW, the `Call` implementation always tries to send a cancellation to the remote peer when its governing context ends -- the trouble for this situation is _we_ may (below) see the context termination before Call does. (Not a disagreement with the comment, just noting that we are only racing between Call notifying us it has done the needful, and our own detection)

It seems like you have this sorted out, but another possible way we could jigger it, I think, would be to [detach](https://pkg.go.dev/context#WithoutCancel) the context we pass to Call from ours, and plumb in one whose cancellation we handle explicitly when we are ready to be done?

Not a blocker though.